sui-execution/latest/sui-move-natives/src/transfer.rs

```diff
@@ -133,14 +156,26 @@ pub fn share_object(
     let cost = context.gas_used();
     Ok(match transfer_result {
         // New means the ID was created in this transaction
-        // SameOwner means the object was previously shared and was re-shared; since
-        // shared objects cannot be taken by-value in the adapter, this can only
-        // happen via test_scenario
+        // SameOwner means the object was previously shared and was re-shared
         TransferResult::New | TransferResult::SameOwner => NativeResult::ok(cost, smallvec![]),
         TransferResult::OwnerChanged => NativeResult::err(cost, E_SHARED_NON_NEW_OBJECT),
     })
 }
 
+fn object_is_shared(context: &mut NativeContext, obj: &Value) -> PartialVMResult<bool> {
```

Comment:
Curious to get thoughts from @tnowacki on the merits of doing this check here vs doing all these consistency checks in the `ObjectRuntime` during `apply_object_changes`?  To my mind:

- Benefits of doing it here is that we stop the transaction as soon as we know it happens, so we save people gas, and unlike wrapped objects, we don't need to wait until the end to be sure.
- Down sides are that the errors from these native functions are quite obtuse, and the implementation is spread out over a number of files.  This function is particular is quite cludgy (Getting the object ID from the

Reply:
We had discussed it since we error in the native function for `transfer::share_object`, and we felt it would be consistent with that behavior to error in those functions. 

However, I agree this implementation feels a bit odd. I would probably modify `TransferResult` and `object_runtime_transfer` to pass back the previous owner to make this all easier.

Reply:
(more context: that was the previous intention of `TransferResult`, to stop this file from having to call into implementation details of the object runtime, and to just stick with one or two functions as an exposed API)

Reply:
Yes, @tnowacki and I spoke briefly about stopping the transaction as soon as we know it will fail when possible, so I think he will be in favor of this

Reply:
If we want to make this change of passing back the owner instead of "New / OwnerChanged / SameOwner" the case when the object is New can't be deduced from outside the object runtime, so we would still have to pass back some flavor of this existing enum.